src/channel/NodeChannel.js

```diff
@@ -18,73 +18,151 @@
+/** @type {{ [key: string]: Client }} */
+const clientCache = {};
+
 export default class NodeChannel extends Channel {
     /**
      * @internal
      * @param {string} address
-     * @param {string=} cert
      * @param {number=} maxExecutionTime
      */
-    constructor(address, cert, maxExecutionTime) {
+    constructor(address, maxExecutionTime) {
         super();
 
-        this.cert = cert;
+        /** @type {Client | null} */
+        this._client = null;
+
+        this.address = address;
         this.maxExecutionTime = maxExecutionTime;
 
+        const { ip, port } = this.parseAddress(address);
+        this.nodeIp = ip;
+        this.nodePort = port;
+    }
+
+    /**
+     * Convert certificate bytes to PEM format
+     * @param {Buffer} certBytes
+     * @returns {string}
+     */
+    bytesToPem(certBytes) {
+        const base64Cert = certBytes.toString("base64");
+        const lines = base64Cert.match(/.{1,64}/g)?.join("\n") || "";
+        return `-----BEGIN CERTIFICATE-----\n${lines}\n-----END CERTIFICATE-----`;
+    }
+
+    /**
+     * Validates and parses an address in the "IP:Port" format.
+     * @param {string} address
+     * @returns {{ ip: string, port: string }}
+     */
+    parseAddress(address) {
+        const [ip, port] = address.split(":");
+        if (!ip || !port) {
+            throw new Error(
+                "Invalid address format. Expected format: 'IP:Port'",
+            );
+        }
+        return { ip, port };
+    }
+
+    /**
+     * Retrieve the server's certificate dynamically.
+     * @param {string} address
+     * @returns {Promise<string>}
+     */
+    async _retrieveCertificate(address) {
+        if (certificateCache[address]) {
+            return certificateCache[address];
+        }
+
+        return new Promise((resolve, reject) => {
+            const socket = tls.connect(
+                {
+                    host: this.nodeIp,
+                    port: Number(this.nodePort),
+                    rejectUnauthorized: false,
+                },
+                () => {
+                    try {
+                        const cert = socket.getPeerCertificate();
+
+                        if (cert && cert.raw) {
+                            const certPEM = this.bytesToPem(cert.raw);
+                            certificateCache[address] = certPEM;
+
+                            resolve(certPEM);
+                        } else {
+                            reject(new Error("No certificate retrieved."));
+                        }
+                    } catch (err) {
+                        reject(err);
+                    } finally {
+                        socket.end();
+                    }
+                },
+            );
+
+            socket.on("error", reject);
+        });
+    }
+
+    /**
+     * Initialize the gRPC client
+     * @returns {Promise<void>}
+     */
+    async _initializeClient() {
+        if (clientCache[this.address]) {
```

Comment:
if i have 1 client executing 2 transactions, the flow will get through here 2 times, right?

Reply:
Yep and since we have caching in place, it will check if the client is already in the cache and return the cached client instead of going through the entire flow again.